test/beast/core/test_helpers.hpp

```diff
@@ -0,0 +1,218 @@
+        std::ostream &os,
+        fcall const &fc)
+    {
+        os << fc.fname << fc.args_ << ';';
+        return os;
+    }
+
+    operator std::string() const
+    {
+        std::ostringstream ss;
+        ss << (*this);
+        return ss.str();
+    }
+
+    const char *fname;
+    args_ref<Args...> args_;
+};
+
+template<class...Args>
+auto
+invocation(
+    const char *fname,
+    Args const &...as) -> fcall<Args...>
+{
+    return fcall<Args...>{fname, args(as...)};
+}
+
+// ====
+
+inline
+std::string
+generate_reference_data(std::size_t length = 2048)
+{
+    std::string result;
+    std::random_device rng;
+    std::seed_seq sequence{rng(), rng(), rng(), rng(), rng()};
+    auto engine = std::default_random_engine(sequence);
+    auto range = std::uniform_int_distribution<unsigned int>(0, 61);
+    std::generate_n(std::back_inserter(result),
+                    length,
+                    [&]{
+                        auto code = range(engine);
+                        if (code < 10)
+                            return '0' + char(code);
+                        else if (code < 36)
+                            return 'A' + char(code - 10);
+                        else
+                            return 'a' + char(code - 36);
+                    });
+    return result;
+}
+
+struct intervals
+{
+    struct iterator
+    {
+        using value_type = std::size_t;
+        using pointer = value_type *;
+        using reference = value_type &;
+        using different_type = std::size_t;
+        using iterator_category = std::input_iterator_tag;
+
+        iterator()
+            : i_(0), n_(0), limit_(0)
+        {}
+
+        iterator(std::size_t i, std::size_t n, std::size_t limit)
+            : i_(i), n_(n), limit_(limit)
+        {}
+
+        bool operator==(iterator const& b) const
+        {
+            auto* ref = b.limit_ == 0 ? this : &b;
+            return i_ == ref->limit_;
+        }
+
+        bool operator!=(iterator const& b) const
+        {
+            return !((*this) == b);
+        }
+
+        iterator& operator++()
+        {
+            i_ += (std::min)(n_, limit_ - i_);
+            return *this;
+        }
+
+        iterator operator++(int)
+        {
+            auto result = *this;
+            ++(*this);
+            return result;
+        }
+
+        value_type operator*() const
+        {
+            return i_;
+        }
+
+    private:
+        std::size_t i_;
+        std::size_t n_;
+        std::size_t limit_;
+    };
+
+    intervals(std::size_t start, std::size_t limit, std::size_t interval)
+        : start_(start)
+        , limit_(limit)
+        , interval_(interval)
+    {}
+
+    auto begin() const -> iterator
+    {
+        return iterator(start_, interval_, limit_);
+    }
+
+    auto end() const -> iterator
+    {
+        return iterator();
+    }
+private:
+    std::size_t start_;
+    std::size_t limit_;
+    std::size_t interval_;
+};
+
+
+} // test_heplers
```

Comment:
No I've decided I don't want to start adding this to every test, we need to think of a better way.